### ISSUE TITLE
Rasa core should use the message text from NLU

### ISSUE THREAD

**Issue by imsaurabh50** [RasaHQ/rasa#5732]
**Description of Problem**:

So when a new message comes it goes through the NLU pipeline for processing. During NLU processing, if I want to change the text of the message due to some reason (say correcting spelling mistakes or anything else.), this corrected text should go to the rasa core for taking further actions. Here is the [forum thread](https://forum.rasa.com/t/alter-user-message-text/26308?u=saurabh-m523) where this was actually asked.

**Overview of the Solution**:

Since rasa core uses the intent and entities identified by NLU for deciding the next actions, it should not use the original text of the message, instead, get the text from NLU. This way the developers have greater flexibility in case if they want to alter the text of the message due to any reason.

**Comment by sara-tagger** [comment]
Thanks for submitting this feature request 🚀 @degiz will get back to you about it soon! ✨

**Comment by degiz** [comment]
Hey @MadPr0grammer 

> say correcting spelling mistakes or anything else

Would it make sense to do that before the message reaches the NLU? Since the messages come from a channel, you can intercept the text (and fix) typos before even sending it further.

Changing the text inside NLU to fix typos sounds a little like not-the-NLU task to me.

**Comment by degiz** [comment]
Hey @MadPr0grammer 

After thinking about it a little longer I've realised that you've mentioned:

>Since rasa core uses the intent and entities identified by NLU for deciding the next actions, it should not use the original text of the message

Rasa Core is prediction next action based on the intents/entities, and doesn't have access to the original text.

Continuing about the stories with typos, I think it makes sense to: 

- include the training examples with typos to the training data.
- or if you want to actually have a `SpellChecker`, you can build a custom NLU component, that should be included in the `config.yml` before the tokenizer.

**Comment by MadPr0grammer** [comment]
Hi @degiz !

The "spell check" was just a simple possible use case. My actual use case is a bit complex:

I want to set some specific slots **along with the user's message**. There is some logic, based on which I'm deciding when to set those slots. This logic-based decision happens at the connector (i.e. before sending it to rasa). But I couldn't use the format: `/intent{"entity1": val1, "entity2": val2}` because I want the NLU to make the intent prediction for me. So, I built my custom NLU component to which I could send messages like `user original text message{"entity1": val1, "entity2": val2}`. This custom connector would then extract the entities from this message and then strip off the entity detection part (i.e. `{"entity1": val1, "entity2": val2}`) from the text and pass the remaining text (i.e. `user original text message`) further down the NLU pipeline for processing. 

Now this works perfectly fine because I have slots of the same name as the entities extracted by my custom component, so rasa core is setting those slots, and I'm able to use slot's value in a custom action.

The problem is that the core is logging the original message text (i.e. `user original text message{"entity1": val1, "entity2": val2}`), instead I want it to log (and use) the message text which comes out of the NLU pipeline after processing (i.e. `user original text message`)

> Rasa Core is prediction next action based on the intents/entities, and doesn't have access to the original text.

It seems that the rasa core does have access to the original text and use it to get the intent and entities in case if the message text is of the format: `/intent{"entity1": val1, "entity2": val2}`. See this [line of the code](https://github.com/RasaHQ/rasa/blob/563f06da7e5a84546c3e2bec661a28f4a3880fdd/rasa/core/processor.py#L459). Please correct me if I'm wrong here.

I hope you understand what I'm trying to say.

**Comment by Samveed** [comment]
Did you find a solution for this? I am facing a similar problem. I want to use a custom NLU component to change the original user message and then, do the next level of processing on the modified message .

**Comment by degiz** [comment]
Hey @MadPr0grammer 

> for its processing

By "the processing" you mean logging it to the tracker right? 

I think if you want the NLU to overwrite the original message, you can easily do that. But I'm not sure that we want to include that to the public code at the moment though. Wdyt?

**Comment by MadPr0grammer** [comment]
Hi @degiz!
Looking at the code, I think (I may be wrong though) that the core checks if the text starts with "/" then it parses the text to get intent and entities, otherwise it uses whatever the NLU has predicted. In any case the core logs the text to the tracker.

So, by processing I mean it should log the text to the tracker and also use the text to determine if it wants to parse the text for getting intent and entities or if it should use what the NLU has predicted.

So in other words, the core should do whatever it does currently, but instead of using the original message, it should use the message that comes out of the NLU pipeline.

PS: Sorry for the late reply 😅
